File: ai/models/bio_age.py
"""Idade biológica, idade fisiológica estimada a partir de literatura revisada.

O que este número É: a idade em que os marcadores medidos desta pessoa seriam
a MEDIANA da população. O que ele NÃO é: um relógio epigenético, um exame ou
qualquer coisa que se pareça com diagnóstico. A tela diz isso, e este módulo
também.

## O caminho da conta

1. **VO2max estimado**, equação de não-exercício de Jurca et al. (2005),
   que usa sexo, idade, IMC, FC de repouso e nível de atividade física. É a
   única das quatro entradas que o AssumFit não precisa pedir: a pulseira mede
   a FC de repouso e o app já conta os minutos de treino da semana.
2. **Idade da aptidão**, a idade cuja mediana de VO2max (registro FRIEND,
   Kaminsky et al. 2015) iguala o VO2max estimado. É o mesmo raciocínio do
   "fitness age" do NTNU, com as normas americanas medidas em esteira.
3. **Idade do HRV**, a relação do RMSSD com a idade é uma lei de potência
   medida por PPG de pulseira em 8 milhões de pessoas (Natarajan et al. 2020),
   e por isso pode ser invertida analiticamente.
4. **Idade do sono profundo**, o N3 cai cerca de 2 pontos percentuais por
   década até os 60 anos (Ohayon et al. 2004).
5. **Combinação**, média ponderada das idades equivalentes que existem, com
   limite por fator e no total.

A versão anterior deste arquivo comparava percentis INVENTADOS (o próprio
arquivo de dados avisava, em letras maiúsculas, que eram "valores
provisórios") e somava pesos em anos escolhidos a olho. O número saía
plausível e não significava nada.

## O que mudou no contrato

`calc_bio_age` passou a aceitar `bmi` e `weekly_active_min`, e a FC de repouso
virou obrigatória de fato, sem ela não há aptidão, que é o eixo principal.
SpO₂ e temperatura saíram do cálculo: não achei norma por idade que suporte
converter qualquer um dos dois em anos, e fator sem fonte é exatamente o que
esta reescrita veio remover. Os dois continuam medidos e exibidos nas telas
deles.

Esta implementação é a fonte da verdade. A cópia em TypeScript no app existe
para funcionar offline e é verificada contra esta por teste de paridade.
"""

from __future__ import annotations

import json
from dataclasses import dataclass
from pathlib import Path
from typing import Literal

Sex = Literal["f", "m"]

_REF = json.loads((Path(__file__).parent.parent / "data" / "bio_age_references.json").read_text())
_VO2_EQ = _REF["vo2max_nao_exercicio"]
_VO2_NORMA = _REF["vo2max_mediana_por_idade"]
_HRV = _REF["hrv_por_idade"]
_SONO = _REF["sono_profundo_por_idade"]
_PESOS = _REF["combinacao"]["pesos"]
_LIM = _REF["limites"]

#: 1 MET = 3,5 mL O2 · kg-1 · min-1 — a definição usada pelo próprio Jurca.
ML_POR_MET = 3.5


@dataclass(frozen=True)
class Factor:
    key: str
    label: str
    value: str
    reference: str
    #: Anos que ESTE fator soma à idade real. Negativo rejuvenesce.
    years: float


@dataclass(frozen=True)
class BioAgeResult:
    real_age: int
    bio_age: int
    #: real_age − bio_age. Positivo = mais jovem que a idade cronológica.
    delta: int
    factors: list[Factor]
    #: VO2max estimado, em mL/kg/min — o número intermediário que mais explica
    #: o resultado, e que a tela mostra por isso.
    vo2max: float | None = None

    def to_dict(self) -> dict:
        return {
            "real_age": self.real_age,
            "bio_age": self.bio_age,
            "delta": self.delta,
            "factors": [f.__dict__ for f in self.factors],
            "vo2max": None if self.vo2max is None else round(self.vo2max, 1),
        }


def _clamp(v: float, lo: float, hi: float) -> float:
    return max(lo, min(hi, v))


def activity_level(weekly_active_min: float | None) -> int:
    """Minutos semanais registrados → categoria SR-PA de Jurca (0 a 4).

    O artigo pergunta por autorrelato; aqui a categoria sai do que foi de fato
    registrado no app, treino do plano concluído mais sessão de esporte. É
    dado medido no lugar de lembrança, e a única liberdade que tomamos com a
    equação original.
    """
    if weekly_active_min is None:
        return 0
    nivel = 0
    for faixa in _VO2_EQ["atividade"]:
        if weekly_active_min >= faixa["min_semanais"]:
            nivel = faixa["nivel"]
    return nivel


def estimate_vo2max(
    *, age: int, sex: Sex, bmi: float, resting_hr: float, weekly_active_min: float | None
) -> float:
    """VO2max em mL/kg/min pela equação de não-exercício (Jurca 2005, NASA)."""
    nivel = activity_level(weekly_active_min)
    coef_pa = next(f["coef"] for f in _VO2_EQ["atividade"] if f["nivel"] == nivel)

    mets = (
        _VO2_EQ["intercepto"]
        + (_VO2_EQ["sexo_masculino"] if sex == "m" else 0.0)
        + _VO2_EQ["por_ano_de_idade"] * age
        + _VO2_EQ["por_ponto_de_imc"] * _clamp(bmi, _LIM["imc_minimo"], _LIM["imc_maximo"])
        + _VO2_EQ["por_bpm_de_fc_repouso"] * resting_hr
        + coef_pa
    )
    # Piso de 1 MET: a equação é linear e pode devolver negativo em combinações
    # extremas (IMC 50 com FC 100), que não existem em fisiologia.
    return max(1.0, mets) * ML_POR_MET


def fitness_age(vo2max: float, sex: Sex) -> float:
    """A idade cuja mediana populacional de VO2max é este valor (FRIEND).

    Interpola linearmente entre os pontos médios das décadas e extrapola pela
    inclinação das pontas, quem tem VO2max acima da mediana dos 20 anos não
    "não tem idade": tem uma idade estimada abaixo de 24,5, e o clamp global
    cuida do resto.
    """
    idades = _VO2_NORMA["idade_central"]
    medianas = _VO2_NORMA[sex]

    # A curva é decrescente: VO2max alto → idade baixa.
    if vo2max >= medianas[0]:
        inclinacao = (medianas[1] - medianas[0]) / (idades[1] - idades[0])
        return idades[0] + (vo2max - medianas[0]) / inclinacao
    if vo2max <= medianas[-1]:
        inclinacao = (medianas[-1] - medianas[-2]) / (idades[-1] - idades[-2])
        return idades[-1] + (vo2max - medianas[-1]) / inclinacao

    for i in range(len(idades) - 1):
        alto, baixo = medianas[i], medianas[i + 1]
        if baixo <= vo2max <= alto:
            fracao = (alto - vo2max) / (alto - baixo)
            return idades[i] + fracao * (idades[i + 1] - idades[i])
    return float(idades[-1])


def hrv_age(rmssd_ms: float, sex: Sex) -> float:
    """A idade em que este RMSSD é o típico (Natarajan 2020, PPG de pulseira).

    Inverte HRV(idade) = HRV30 · (idade/30)^α. Como α é negativo, RMSSD maior
    devolve idade menor, que é o comportamento fisiológico esperado.
    """
    p = _HRV[sex]
    if rmssd_ms <= 0:
        return float(_HRV["idade_de_referencia"])
    return _HRV["idade_de_referencia"] * (rmssd_ms / p["rmssd_aos_30"]) ** (1.0 / p["alfa"])


def deep_sleep_age(deep_fraction: float) -> float:
    """A idade em que esta fração de sono profundo é a esperada (Ohayon 2004).

    A relação vale até os 60 anos; depois o N3 estabiliza. Acima do platô a
    função devolve a própria idade de platô, porque ali o marcador deixa de
    discriminar idade, afirmar 80 anos a partir de N3 baixo seria ler a
    curva onde ela não existe.
    """
    esperado_no_plato = _SONO["fracao_aos_30"] - _SONO["queda_por_ano"] * (
        _SONO["idade_de_platô"] - 30
    )
    if deep_fraction <= esperado_no_plato:
        return float(_SONO["idade_de_platô"])
    return 30 + (_SONO["fracao_aos_30"] - deep_fraction) / _SONO["queda_por_ano"]


def calc_bio_age(
    *,
    real_age: int,
    sex: Sex,
    hrv_ms: float | None,
    resting_hr: float,
    deep_sleep_pct: float | None,
    bmi: float | None = None,
    weekly_active_min: float | None = None,
    # Aceitos e IGNORADOS: os dois entravam no cálculo antigo com peso
    # inventado. Continuam no contrato para não quebrar quem chama, e o
    # docstring do módulo explica por que saíram da conta.
    spo2_pct: float | None = None,
    temp_range_c: float | None = None,
) -> BioAgeResult:
    imc = _clamp(bmi if bmi is not None else _LIM["imc_padrao"], _LIM["imc_minimo"], _LIM["imc_maximo"])
    limite = _LIM["desvio_maximo_por_fator_anos"]

    vo2 = estimate_vo2max(
        age=real_age, sex=sex, bmi=imc, resting_hr=resting_hr, weekly_active_min=weekly_active_min
    )
    # Dois limites, por razões diferentes: o DOMÍNIO é onde o estudo olhou
    # (20 a 79 anos) e fora dele não se afirma nada; o limite por fator protege
    # contra artefato de medição, que é outro problema.
    piso, teto = _VO2_NORMA["idade_minima_afirmavel"], _VO2_NORMA["idade_maxima_afirmavel"]
    idade_aptidao = _clamp(
        _clamp(fitness_age(vo2, sex), piso, teto), real_age - limite, real_age + limite
    )

    idade_hrv = (
        None
        if hrv_ms is None
        else _clamp(
            _clamp(hrv_age(hrv_ms, sex), _HRV["idade_minima_afirmavel"], _HRV["idade_maxima_afirmavel"]),
            real_age - limite,
            real_age + limite,
        )
    )
    idade_sono = (
        None
        if deep_sleep_pct is None
        else _clamp(deep_sleep_age(deep_sleep_pct), real_age - limite, real_age + limite)
    )

    # Média ponderada só do que existe: o peso do ausente é redistribuído entre
    # os presentes, nunca tratado como zero.
    partes = [(idade_aptidao, _PESOS["aptidao"])]
    if idade_hrv is not None:
        partes.append((idade_hrv, _PESOS["hrv"]))
    if idade_sono is not None:
        partes.append((idade_sono, _PESOS["sono"]))
    peso_total = sum(p for _, p in partes)
    idade_estimada = sum(v * p for v, p in partes) / peso_total

    desvio = _clamp(
        idade_estimada - real_age,
        -_LIM["desvio_maximo_total_anos"],
        _LIM["desvio_maximo_total_anos"],
    )
    bio_age = max(_LIM["idade_minima"], round(real_age + desvio))

    nivel = activity_level(weekly_active_min)
    descricao_atividade = next(f["descricao"] for f in _VO2_EQ["atividade"] if f["nivel"] == nivel)

    factors = [
        Factor(
            "fitness",
            "Aptidão cardiorrespiratória",
            f"VO₂máx {vo2:.1f} ml/kg/min",
            f"mediana da sua idade: {_mediana_na_idade(real_age, sex):.1f}",
            idade_aptidao - real_age,
        ),
        Factor(
            "hrv",
            "HRV",
            "–" if hrv_ms is None else f"{round(hrv_ms)} ms",
            f"típico aos {real_age}: {_hrv_tipico(real_age, sex):.0f} ms",
            0.0 if idade_hrv is None else idade_hrv - real_age,
        ),
        Factor(
            "sleep",
            "Sono profundo",
            "–" if deep_sleep_pct is None else f"{round(deep_sleep_pct * 100)}%",
            f"típico aos {real_age}: {_sono_tipico(real_age) * 100:.0f}%",
            0.0 if idade_sono is None else idade_sono - real_age,
        ),
        Factor(
            "activity",
            "Atividade semanal",
            "–" if weekly_active_min is None else f"{round(weekly_active_min)} min",
            descricao_atividade,
            # A atividade não é um fator à parte: ela ENTRA na aptidão, e
            # somá-la de novo seria contá-la duas vezes.
            0.0,
        ),
    ]

    return BioAgeResult(
        real_age=real_age,
        bio_age=bio_age,
        delta=real_age - bio_age,
        factors=factors,
        vo2max=vo2,
    )


def _mediana_na_idade(age: int, sex: Sex) -> float:
    """A mediana de VO2max da idade da pessoa, a régua contra a qual ela é lida."""
    idades = _VO2_NORMA["idade_central"]
    medianas = _VO2_NORMA[sex]
    if age <= idades[0]:
        return medianas[0]
    if age >= idades[-1]:
        return medianas[-1]
    for i in range(len(idades) - 1):
        if idades[i] <= age <= idades[i + 1]:
            fracao = (age - idades[i]) / (idades[i + 1] - idades[i])
            return medianas[i] + fracao * (medianas[i + 1] - medianas[i])
    return medianas[-1]


def _hrv_tipico(age: int, sex: Sex) -> float:
    p = _HRV[sex]
    return p["rmssd_aos_30"] * (age / _HRV["idade_de_referencia"]) ** p["alfa"]


def _sono_tipico(age: int) -> float:
    idade = min(age, _SONO["idade_de_platô"])
    return _SONO["fracao_aos_30"] - _SONO["queda_por_ano"] * (idade - 30)

File: app/src/domain/bioAge.ts
/**
 * Idade biológica — idade fisiológica estimada, a partir de literatura revisada.
 *
 * Espelho em TypeScript de `ai/models/bio_age.py`, que é a fonte da verdade.
 * Existe para o número aparecer offline, e o teste de paridade
 * (`ai/tests/test_models.py::TestParidadeComTypeScript`) impede as duas contas
 * de divergirem em silêncio. **Mexeu aqui, rode aquele teste.**
 *
 * O que este número é: a idade em que os marcadores medidos seriam a MEDIANA
 * da população. Não é relógio epigenético, exame nem diagnóstico.
 *
 * O caminho da conta, com as fontes:
 *
 * 1. VO₂máx estimado por equação de não-exercício — Jurca R, et al. *Assessing
 *    cardiorespiratory fitness without performing exercise testing*. Am J Prev
 *    Med. 2005;29(3):185-193 (Tabela 5, coluna NASA).
 * 2. Idade da aptidão pela mediana populacional — Kaminsky LA, Arena R, Myers
 *    J. Mayo Clin Proc. 2015;90(11):1515-1523 (registro FRIEND, esteira).
 * 3. Idade do HRV pela lei de potência do RMSSD medido por PPG de pulseira —
 *    Natarajan A, et al. Lancet Digit Health. 2020;2(12):e650-e657 (Tabela 1).
 * 4. Idade do sono profundo — Ohayon MM, et al. Sleep. 2004;27(7):1255-1273.
 *
 * A versão anterior comparava percentis inventados (o arquivo de dados avisava
 * disso em maiúsculas) e somava pesos em anos escolhidos a olho: o número saía
 * plausível e não significava nada.
 */
import { DASH } from './ratings';
import type { BioAge, BioAgeFactor, Sex } from './types';

/** 1 MET = 3,5 mL O₂ · kg⁻¹ · min⁻¹ — a definição do próprio Jurca. */
const ML_POR_MET = 3.5;

/** Jurca 2005, Tabela 5 (NASA): R = 0,81; SEE = 1,45 MET. */
const VO2_EQ = {
  intercepto: 18.07,
  sexoMasculino: 2.77,
  porAnoDeIdade: -0.1,
  porPontoDeImc: -0.17,
  porBpmDeFcRepouso: -0.03,
};

/**
 * Categorias SR-PA de Jurca (Tabela 1, NASA), traduzidas para os minutos
 * semanais REGISTRADOS no app — dado medido no lugar de autorrelato.
 */
const ATIVIDADE = [
  { nivel: 0, coef: 0.0, minSemanais: 0, descricao: 'pouca atividade além de caminhar' },
  { nivel: 1, coef: 0.32, minSemanais: 1, descricao: 'alguma prática regular de esporte ou recreação' },
  { nivel: 2, coef: 1.06, minSemanais: 20, descricao: '20 a 60 minutos de aeróbio por semana' },
  { nivel: 3, coef: 1.76, minSemanais: 60, descricao: '1 a 3 horas de aeróbio por semana' },
  { nivel: 4, coef: 3.03, minSemanais: 180, descricao: 'mais de 3 horas de aeróbio por semana' },
];

/** FRIEND (Kaminsky 2015): percentil 50 de VO₂máx no ponto médio de cada década. */
const VO2_NORMA = {
  idadeCentral: [24.5, 34.5, 44.5, 54.5, 64.5, 74.5],
  m: [48.0, 42.4, 37.8, 32.6, 28.2, 24.4],
  f: [37.6, 30.2, 26.7, 23.4, 20.0, 18.3],
  /** O estudo cobre 20 a 79 anos; fora disso não se afirma idade. */
  idadeMinima: 20,
  idadeMaxima: 79,
};

/** Natarajan 2020, Tabela 1: HRV(idade) = HRV30 × (idade/30)^α. */
const HRV_NORMA = {
  idadeDeReferencia: 30,
  m: { rmssdAos30: 44.8, alfa: -0.804 },
  f: { rmssdAos30: 43.7, alfa: -0.666 },
  idadeMinima: 20,
  idadeMaxima: 79,
};

/** Ohayon 2004: o N3 cai ~2 pontos percentuais por década até os 60 anos. */
const SONO_NORMA = { fracaoAos30: 0.2, quedaPorAno: 0.002, idadeDePlato: 60 };

/**
 * Peso de cada idade equivalente. É CALIBRAÇÃO declarada, não resultado de
 * estudo: a aptidão pesa mais porque tem a associação mais forte com
 * mortalidade entre os marcadores usados (o próprio Jurca cita 9–10 METs em
 * homens e 7–8 em mulheres associados a ≥50% menos risco de morte).
 */
const PESOS = { aptidao: 0.6, hrv: 0.25, sono: 0.15 };

const LIMITES = {
  desvioMaximoPorFator: 20,
  desvioMaximoTotal: 15,
  idadeMinima: 18,
  imcMinimo: 15,
  imcMaximo: 50,
  /** Meio da faixa saudável da OMS: o palpite que menos move o resultado. */
  imcPadrao: 24,
};

const clamp = (v: number, lo: number, hi: number) => Math.max(lo, Math.min(hi, v));

export type BioAgeInput = {
  realAge: number;
  sex: Sex;
  hrvMs: number | null;
  restingHr: number;
  /** `null` sem noite medida — sai da média em vez de valer zero. */
  deepSleepPct: number | null;
  /** Da anamnese (peso e altura). `null` usa o IMC padrão. */
  bmi?: number | null;
  /** Minutos de treino e esporte registrados na semana. */
  weeklyActiveMin?: number | null;
  /** Aceitos e ignorados: saíram do cálculo por falta de norma por idade. */
  spo2Pct?: number | null;
  tempRangeC?: number | null;
};

export function activityLevel(weeklyActiveMin: number | null | undefined): number {
  if (weeklyActiveMin == null) return 0;
  let nivel = 0;
  for (const faixa of ATIVIDADE) if (weeklyActiveMin >= faixa.minSemanais) nivel = faixa.nivel;
  return nivel;
}

export function estimateVo2max(input: {
  age: number;
  sex: Sex;
  bmi: number;
  restingHr: number;
  weeklyActiveMin: number | null | undefined;
}): number {
  const nivel = activityLevel(input.weeklyActiveMin);
  const coefPa = ATIVIDADE.find((f) => f.nivel === nivel)!.coef;

  const mets =
    VO2_EQ.intercepto +
    (input.sex === 'm' ? VO2_EQ.sexoMasculino : 0) +
    VO2_EQ.porAnoDeIdade * input.age +
    VO2_EQ.porPontoDeImc * clamp(input.bmi, LIMITES.imcMinimo, LIMITES.imcMaximo) +
    VO2_EQ.porBpmDeFcRepouso * input.restingHr +
    coefPa;

  // Piso de 1 MET: a reta pode devolver negativo em combinações extremas que
  // não existem em fisiologia.
  return Math.max(1, mets) * ML_POR_MET;
}

/** A idade cuja mediana populacional de VO₂máx é este valor (FRIEND). */
export function fitnessAge(vo2max: number, sex: Sex): number {
  const idades = VO2_NORMA.idadeCentral;
  const medianas = VO2_NORMA[sex];

  if (vo2max >= medianas[0]) {
    const inclinacao = (medianas[1] - medianas[0]) / (idades[1] - idades[0]);
    return idades[0] + (vo2max - medianas[0]) / inclinacao;
  }
  if (vo2max <= medianas[medianas.length - 1]) {
    const n = idades.length - 1;
    const inclinacao = (medianas[n] - medianas[n - 1]) / (idades[n] - idades[n - 1]);
    return idades[n] + (vo2max - medianas[n]) / inclinacao;
  }
  for (let i = 0; i < idades.length - 1; i++) {
    const alto = medianas[i];
    const baixo = medianas[i + 1];
    if (vo2max <= alto && vo2max >= baixo) {
      const fracao = (alto - vo2max) / (alto - baixo);
      return idades[i] + fracao * (idades[i + 1] - idades[i]);
    }
  }
  return idades[idades.length - 1];
}

/** A idade em que este RMSSD é o típico (Natarajan 2020). */
export function hrvAge(rmssdMs: number, sex: Sex): number {
  const p = HRV_NORMA[sex];
  if (rmssdMs <= 0) return HRV_NORMA.idadeDeReferencia;
  return HRV_NORMA.idadeDeReferencia * Math.pow(rmssdMs / p.rmssdAos30, 1 / p.alfa);
}

/** A idade em que esta fração de sono profundo é a esperada (Ohayon 2004). */
export function deepSleepAge(deepFraction: number): number {
  const esperadoNoPlato =
    SONO_NORMA.fracaoAos30 - SONO_NORMA.quedaPorAno * (SONO_NORMA.idadeDePlato - 30);
  if (deepFraction <= esperadoNoPlato) return SONO_NORMA.idadeDePlato;
  return 30 + (SONO_NORMA.fracaoAos30 - deepFraction) / SONO_NORMA.quedaPorAno;
}

function medianaNaIdade(age: number, sex: Sex): number {
  const idades = VO2_NORMA.idadeCentral;
  const medianas = VO2_NORMA[sex];
  if (age <= idades[0]) return medianas[0];
  if (age >= idades[idades.length - 1]) return medianas[medianas.length - 1];
  for (let i = 0; i < idades.length - 1; i++) {
    if (age >= idades[i] && age <= idades[i + 1]) {
      const fracao = (age - idades[i]) / (idades[i + 1] - idades[i]);
      return medianas[i] + fracao * (medianas[i + 1] - medianas[i]);
    }
  }
  return medianas[medianas.length - 1];
}

const hrvTipico = (age: number, sex: Sex) =>
  HRV_NORMA[sex].rmssdAos30 * Math.pow(age / HRV_NORMA.idadeDeReferencia, HRV_NORMA[sex].alfa);

const sonoTipico = (age: number) =>
  SONO_NORMA.fracaoAos30 - SONO_NORMA.quedaPorAno * (Math.min(age, SONO_NORMA.idadeDePlato) - 30);

export function calcBioAge(input: BioAgeInput): BioAge {
  const { realAge, sex, hrvMs, restingHr, deepSleepPct } = input;
  const imc = clamp(input.bmi ?? LIMITES.imcPadrao, LIMITES.imcMinimo, LIMITES.imcMaximo);
  const limite = LIMITES.desvioMaximoPorFator;

  const vo2 = estimateVo2max({
    age: realAge,
    sex,
    bmi: imc,
    restingHr,
    weeklyActiveMin: input.weeklyActiveMin,
  });

  // Dois limites, por razões diferentes: o DOMÍNIO é onde o estudo olhou, e o
  // limite por fator protege contra artefato de medição.
  const idadeAptidao = clamp(
    clamp(fitnessAge(vo2, sex), VO2_NORMA.idadeMinima, VO2_NORMA.idadeMaxima),
    realAge - limite,
    realAge + limite,
  );
  const idadeHrv =
    hrvMs == null
      ? null
      : clamp(
          clamp(hrvAge(hrvMs, sex), HRV_NORMA.idadeMinima, HRV_NORMA.idadeMaxima),
          realAge - limite,
          realAge + limite,
        );
  const idadeSono =
    deepSleepPct == null
      ? null
      : clamp(deepSleepAge(deepSleepPct), realAge - limite, realAge + limite);

  // Média ponderada só do que existe: o peso do ausente é redistribuído.
  const partes: [number, number][] = [[idadeAptidao, PESOS.aptidao]];
  if (idadeHrv != null) partes.push([idadeHrv, PESOS.hrv]);
  if (idadeSono != null) partes.push([idadeSono, PESOS.sono]);
  const pesoTotal = partes.reduce((s, [, p]) => s + p, 0);
  const idadeEstimada = partes.reduce((s, [v, p]) => s + v * p, 0) / pesoTotal;

  const desvio = clamp(
    idadeEstimada - realAge,
    -LIMITES.desvioMaximoTotal,
    LIMITES.desvioMaximoTotal,
  );
  const bioAge = Math.max(LIMITES.idadeMinima, Math.round(realAge + desvio));

  const nivel = activityLevel(input.weeklyActiveMin);
  const descricaoAtividade = ATIVIDADE.find((f) => f.nivel === nivel)!.descricao;

  const factors: BioAgeFactor[] = [
    {
      key: 'fitness',
      label: 'Aptidão cardiorrespiratória',
      value: `VO₂máx ${vo2.toFixed(1).replace('.', ',')} ml/kg/min`,
      reference: `mediana da sua idade: ${medianaNaIdade(realAge, sex).toFixed(1).replace('.', ',')}`,
      years: idadeAptidao - realAge,
    },
    {
      key: 'hrv',
      label: 'HRV',
      value: hrvMs == null ? DASH : `${Math.round(hrvMs)} ms`,
      reference: `típico aos ${realAge}: ${Math.round(hrvTipico(realAge, sex))} ms`,
      years: idadeHrv == null ? 0 : idadeHrv - realAge,
    },
    {
      key: 'sleep',
      label: 'Sono profundo',
      value: deepSleepPct == null ? DASH : `${Math.round(deepSleepPct * 100)}%`,
      reference: `típico aos ${realAge}: ${Math.round(sonoTipico(realAge) * 100)}%`,
      years: idadeSono == null ? 0 : idadeSono - realAge,
    },
    {
      key: 'activity',
      label: 'Atividade semanal',
      value: input.weeklyActiveMin == null ? DASH : `${Math.round(input.weeklyActiveMin)} min`,
      reference: descricaoAtividade,
      // A atividade ENTRA na aptidão; somá-la de novo seria contar duas vezes.
      years: 0,
    },
  ];

  return { realAge, bioAge, delta: realAge - bioAge, factors, vo2max: Number(vo2.toFixed(1)) };
}

/**
 * `-2.4` → `−2,4a`. O sinal é o da CONTA, não da leitura: anos negativos
 * significam que o marcador puxa a idade para baixo — mais jovem.
 */
export function formatYears(years: number): string {
  const sinal = years <= 0 ? '−' : '+';
  return `${sinal}${Math.abs(years).toFixed(1).replace('.', ',')}a`;
}
